contrib/mesos/pkg/executor/service/podsource.go

```diff
@@ -0,0 +1,132 @@
+	// @see ConfigSourceAnnotationKey
+	mesosSource = kubetypes.ApiserverSource
+)
+
+// sourceMesos merges pods from mesos, and mirror pods from the apiserver. why?
+// (a) can't have two sources with the same name;
+// (b) all sources, other than ApiserverSource are considered static/mirror
+// sources, and;
+// (c) kubelet wants to see mirror pods reflected in a non-static source.
+//
+// Mesos pods must appear to come from apiserver due to (b), while reflected
+// static pods (mirror pods) must appear to come from apiserver due to (c).
+//
+// The only option I could think of was creating a source that merges the pod
+// streams. I don't like it. But I could think of anything else, other than
+// starting to hack up the kubelet's understanding of mirror/static pod
+// sources (ouch!)
+type sourceMesos struct {
+	sourceFinished chan struct{}              // sourceFinished closes when mergeAndForward exits
+	out            chan<- interface{}         // out is the sink for merged pod snapshots
+	mirrorPods     chan []*api.Pod            // mirrorPods communicates snapshots of the current set of mirror pods
+	execUpdates    <-chan kubetypes.PodUpdate // execUpdates receives snapshots of the current set of mesos pods
+}
+
+// newSourceMesos creates a pod config source that merges pod updates from
+// mesos (via execUpdates), and mirror pod updates from the apiserver (via
+// podWatch) writing the merged update stream to the out chan. It is expected
+// that execUpdates will only ever contain SET operations. The source takes
+// ownership of the sourceFinished chan, closing it when the source terminates.
+// Source termination happens when the execUpdates chan is closed and fully
+// drained of updates.
+func newSourceMesos(
+	sourceFinished chan struct{},
+	execUpdates <-chan kubetypes.PodUpdate,
+	out chan<- interface{},
+	podWatch *cache.ListWatch,
+) {
+	source := &sourceMesos{
+		sourceFinished: sourceFinished,
+		mirrorPods:     make(chan []*api.Pod),
+		execUpdates:    execUpdates,
+		out:            out,
+	}
+	// reflect changes from the watch into a chan, filtered to include only mirror pods (have an ConfigMirrorAnnotationKey attr)
+	cache.NewReflector(podWatch, &api.Pod{}, cache.NewUndeltaStore(source.send, cache.MetaNamespaceKeyFunc), 0).RunUntil(sourceFinished)
+	go source.mergeAndForward()
+}
+
+func (source *sourceMesos) send(objs []interface{}) {
+	var pods []*api.Pod
+	for _, o := range objs {
+		p := o.(*api.Pod)
+		if _, ok := p.Annotations[kubetypes.ConfigMirrorAnnotationKey]; ok {
+			pods = append(pods, p)
+		}
+	}
+	select {
+	case <-source.sourceFinished:
+	case source.mirrorPods <- pods:
+	}
+}
+
+func (source *sourceMesos) mergeAndForward() {
+	// execUpdates will be closed by the executor on shutdown
+	defer close(source.sourceFinished)
+	var (
+		mirrors = []*api.Pod{}
+		pods    = []*api.Pod{}
+	)
+eventLoop:
+	for {
+		select {
+		case m := <-source.mirrorPods:
+			mirrors = make([]*api.Pod, len(m))
+			copy(mirrors, m)
```

Comment:
I was debugging something and decided to `copy()` instead of grabbing a shallow copy (via `m[:]`)